autopilot/migrator/migrator.go

```diff
@@ -0,0 +1,364 @@
+		UpdateAccounts(context.Context, []api.Account) error
+		UpdateSlab(ctx context.Context, key object.EncryptionKey, sectors []api.UploadedSector) error
+		UploadParams(ctx context.Context) (api.UploadParams, error)
+		UsableHosts(ctx context.Context) (hosts []api.HostInfo, err error)
+	}
+
+	Migrator interface {
+		Migrate(ctx context.Context)
+		SignalMaintenanceFinished()
+		Status() (bool, time.Time)
+		Stop()
+	}
+
+	SlabStore interface {
+		RefreshHealth(ctx context.Context) error
+		Slab(ctx context.Context, key object.EncryptionKey) (object.Slab, error)
+		SlabsForMigration(ctx context.Context, healthCutoff float64, limit int) ([]api.UnhealthySlab, error)
+	}
+)
+
+type (
+	migrator struct {
+		alerts alerts.Alerter
+		bus    Bus
+		ss     SlabStore
+
+		healthCutoff           float64
+		parallelSlabsPerWorker uint64
+
+		masterKey utils.MasterKey
+
+		accounts        *accounts.Manager
+		downloadManager *download.Manager
+		uploadManager   *upload.Manager
+		hostManager     hosts.Manager
+
+		rhp4Client *rhp4.Client
+
+		signalConsensusNotSynced  chan struct{}
+		signalMaintenanceFinished chan struct{}
+
+		statsSlabMigrationSpeedMS *utils.DataPoints
+
+		shutdownCtx context.Context
+		wg          sync.WaitGroup
+
+		logger *zap.SugaredLogger
+
+		mu                 sync.Mutex
+		migrating          bool
+		migratingLastStart time.Time
+	}
+)
+
+func New(ctx context.Context, cfg config.Migrator, masterKey utils.MasterKey, alerts alerts.Alerter, ss SlabStore, b Bus, logger *zap.Logger) (*migrator, error) {
+	logger = logger.Named("migrator")
+
+	if cfg.AccountsRefillInterval == 0 {
+		return nil, fmt.Errorf("accounts refill interval must be set")
+	}
+
+	m := &migrator{
+		alerts: alerts,
+		bus:    b,
+		ss:     ss,
+
+		healthCutoff:           cfg.HealthCutoff,
+		parallelSlabsPerWorker: cfg.ParallelSlabsPerWorker,
+
+		masterKey: masterKey,
+
+		signalConsensusNotSynced:  make(chan struct{}, 1),
+		signalMaintenanceFinished: make(chan struct{}, 1),
+
+		statsSlabMigrationSpeedMS: utils.NewDataPoints(time.Hour),
+
+		shutdownCtx: ctx,
+
+		logger: logger.Sugar(),
+	}
+
+	// create account manager
+	am, err := accounts.NewManager(masterKey.DeriveAccountsKey("migrator"), "migrator", alerts, m, m, b, b, b, b, cfg.AccountsRefillInterval, logger)
+	if err != nil {
+		return nil, err
+	}
+	m.accounts = am
+
+	// create host manager
+	dialer := rhp.NewFallbackDialer(b, net.Dialer{}, logger)
+	csr := contracts.NewSpendingRecorder(ctx, b, 5*time.Second, logger)
+	hm := hosts.NewManager(masterKey, am, csr, dialer, logger)
+	m.hostManager = hm
+	m.rhp4Client = rhp4.New(dialer)
+
+	// create upload & download manager
+	mm := memory.NewManager(math.MaxInt64, logger)
+	uk := masterKey.DeriveUploadKey()
+	m.downloadManager = download.NewManager(ctx, &uk, hm, mm, b, cfg.UploadMaxOverdrive, cfg.UploadOverdriveTimeout, logger)
+	m.uploadManager = upload.NewManager(ctx, &uk, hm, mm, b, b, b, cfg.UploadMaxOverdrive, cfg.UploadOverdriveTimeout, logger)
```

Comment:
All that dependencies the migrator needs should be passed in and not be created by it to keep the dependencies as small as possible.

Reply:
closing this in favour of https://github.com/SiaFoundation/renterd/pull/1735#discussion_r1879824826